### PR TITLE
Add logging 2.1 to 2.2/3.0 migration guide

### LINKED ISSUE #9490: ASP.NET Core 2.2 to 3.0 migration

### PR DESCRIPTION
**Pull request by pakrym**
Fixes: https://github.com/aspnet/Docs/issues/9841
Fixes #9490

### CODE REVIEW

**Comment by pakrym** [comment]
cc @ajcvickers

**Comment by guardrex** [comment]
@pakrym We usually use our in-topic versioning (in this case, in the [Logging topic](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging/)) to cover updates to the framework and then call out the migration steps in the migration topic (in this case, [Migrate from ASP.NET Core 2.1 to 2.2](https://docs.microsoft.com/en-us/aspnet/core/migration/21-to-22)). Idk what they'll suggest, but the pattern has been not to have migration topics in the *Fundamentals* node. Ping Sir @tdykstra for help.

**Comment by tdykstra** [comment]
That's right, this should go in https://github.com/aspnet/Docs/blob/master/aspnetcore/migration/21-to-22.md.  We don't have a place for the 3.0 instructions yet, but they will end up in a 3.0 migration doc located in the same folder.

**Comment by pakrym** [comment]
Should I make two copies of this doc then?

**Comment by tdykstra** [comment]
There's an outstanding issue for writing logging docs that aren't ASP.NET Core-specific -- 

https://github.com/dotnet/docs/issues/4567

Until that is resolved I would put the 2.1 to 2.2 content in the 2.1 to 2.2 doc, and publish the non-ASP.NET-Core content and the 3.0 content in a blog for now. 

Considering the dialog in https://github.com/aspnet/Logging/pull/932, I would also link from the logging doc to the new 2.1 to 2.2 logging section of the migration doc to make that more discoverable.

cc @Rick-Anderson in cases he has any other ideas about where to put non-ASP.NET Core and migrate-to-3.0 content.

**Comment by Rick-Anderson** [comment]
@tdykstra 	can you help move the content around for @pakrym 	in exchange for him helping use on technical content?

**Comment by tdykstra** [comment]
What about blogging the 3.0 and non-ASP.NET parts?

**Comment by pakrym** [comment]
> What about blogging the 3.0 and non-ASP.NET parts?

I think they might need to be referenced from EF logging doc (https://docs.microsoft.com/en-us/ef/core/miscellaneous/logging)

**Comment by tdykstra** [comment]
We could add the non-ASP.NET Core content there until we get a permanent home for it, that leaves 3.0.  Maybe just start a 2.2-to-3.0 migration doc that only has this logging content in it for now?

**Comment by tdykstra** [comment]
So if this plan makes sense, I'll
* Move the 2.1 to 2.2 content to the appropriate migration doc
* Move the 3.0 content to a new (very small) 2.2 to 3.0 migration doc
* Move the non-ASP.NET Core content to the EF logging doc

**Comment by pakrym** [comment]
> Move the non-ASP.NET Core content to the EF logging doc

I'm not a fan of this part, while it's non-ASP.NET Core content it's even less EF specific. I would prefer having it in ASP.NET Core migration doc over EF.

**Comment by guardrex** [comment]
@tdykstra Caught my :eye: looking at 3.0 issues ...

> Move the 3.0 content to a new (very small) 2.2 to 3.0 migration doc

https://github.com/aspnet/Docs/issues/9490 ... content that might make a good basis for it ... and potentially helpful for very *VERY* early adopters playing with the bits.

**Comment by Rick-Anderson** [comment]
@tdykstra 	can we S&M this? You can move this to *a new (very small) 2.2 to 3.0 migration doc* after.

**Comment by tdykstra** [comment]
Better to avoid that, it creates a new URL which we would then have to redirect and we wouldn't have a single logical target to redirect to.  I'll take care of it tomorrow.

**Comment by Rick-Anderson** [comment]
But don't we eventually need a 2.2 to 3.0 migration doc?

**Comment by tdykstra** [comment]
Not in the fundamentals/logging folder

**Comment by Rick-Anderson** [comment]
In the migrations folder.

**Comment by tdykstra** [comment]
Once I realized the only 3.0 content was for non-ASP.NET Core apps, it no longer made sense to put it in a new ASP.NET Core migration doc, so I ended up with this:

* 2.1 to 2.2 ASP.NET Core logging content --> ASP.NET Core 2.1 to 2.2 migration doc
* 2.1 to 2.2 or 3.0 non-ASP.NET Core logging content --> new doc in migration folder, a temporary location until we have a better place for non-ASP.NET Core logging docs.

I had already created the new 2.2 to 3.0 migration doc for #9490 before deciding the logging content didn't fit, so I'm including it as well.

File: aspnetcore/migration/22-to-30.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  </ItemGroup>
+  ```
+
+* Update the `Version` on remaining `<PackageReference>` elements for `Microsoft.AspNetCore.*` packages to 3.0.0-preview-18579-0053 (or whatever the number is for the current Preview)
+
+  If there is no 3.0 version of a package, the package might have been deprecated in 3.0. Many of these are part of `Microsoft.AspNetCore.App` and don't need to be referenced individually anymore. For a preliminary list of packages no longer produced in 3.0, see https://github.com/aspnet/AspNetCore/issues/3756.
+
+* Some APIs were removed from Microsoft.AspNetCore.App between 2.x and 3.0. You may need to add `<PackageReference>` items if you're using API from packages listed in https://github.com/aspnet/AspNetCore/issues/3755
+
+  For example, Entity Framework Core and System.Data.SqlClient are no longer part of Microsoft.AspNetCore.App.
```
[the remark this answers is not in the document]
Comment (Rick-Anderson) [reply aspnetcore/migration/22-to-30.md]: @tdykstra 	can you review my suggestions and apply as appropriate?

File: aspnetcore/migration/22-to-30.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+  ```xml
+  <ItemGroup>
+     <FrameworkReference Include="Microsoft.AspNetCore.App" />
+  </ItemGroup>
+  ```
+
+* Update the `Version` on remaining `<PackageReference>` elements for `Microsoft.AspNetCore.*` packages to 3.0.0-preview-18579-0053 (or whatever the number is for the current Preview)
+
+  If there is no 3.0 version of a package, the package might have been deprecated in 3.0. Many of these are part of `Microsoft.AspNetCore.App` and don't need to be referenced individually anymore. For a preliminary list of packages no longer produced in 3.0, see https://github.com/aspnet/AspNetCore/issues/3756.
```
[the remark this answers is not in the document]
Comment (tdykstra) [reply aspnetcore/migration/22-to-30.md]: > Is there a list of deprecated packages in 3.0?

https://github.com/aspnet/AspNetCore/issues/3756

File: aspnetcore/includes/net-core-prereqs-vs-3.0.md [hunk 3]
```diff
@@ -0,0 +1,2 @@
+* [Visual Studio 2017 version 15.9 or later](https://visualstudio.microsoft.com/downloads/) with the **ASP.NET and web development** workload
```
Comment (scottaddie) [inline aspnetcore/includes/net-core-prereqs-vs-3.0.md]: Use the campaign link to allow tracking:

```suggestion
* [Visual Studio 2017 version 15.9 or later](https://www.visualstudio.com/downloads/?utm_medium=microsoft&utm_source=docs.microsoft.com&utm_campaign=button+cta&utm_content=download+vs2017) with the **ASP.NET and web development** workload
```

Also, I suspect VS 2019 will be required when 3.0 releases. Do we have an issue to update prereqs accordingly?

File: aspnetcore/migration/22-to-30.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+* Some APIs were removed from Microsoft.AspNetCore.App between 2.x and 3.0. You may need to add `<PackageReference>` items if you're using API from packages listed in https://github.com/aspnet/AspNetCore/issues/3755
```
[the remark this answers is not in the document]
Comment (guardrex) [reply aspnetcore/migration/22-to-30.md]: I always include the title in these types of issue links. Suggestion (this is what I use) ...

> \[Assemblies being removed from Microsoft.AspNetCore.App 3.0 (aspnet/AspNetCore #3755)](https:\/\/github.com/aspnet/AspNetCore/issues/3755)

File: aspnetcore/includes/net-core-prereqs-vs-3.0.md [hunk 5]
```diff
@@ -0,0 +1,2 @@
...
+* [.NET Core SDK 3.0 Preview](https://www.microsoft.com/net/download/all)
```
Comment (scottaddie) [inline aspnetcore/includes/net-core-prereqs-vs-3.0.md]: Consider future-proofing this text by dropping the "Preview" moniker:

```suggestion
* [.NET Core SDK 3.0](https://www.microsoft.com/net/download/all)
```

File: aspnetcore/migration/22-to-30.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (guardrex) [reply aspnetcore/migration/22-to-30.md]: ... and we pretty much always link to our metapackage topic (with the word "metapackage" in the link) ...

> \[Microsoft.AspNetCore.App metapackage](xref:fundamentals/metapackage-app)

File: aspnetcore/migration/22-to-30.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+  For example, Entity Framework Core and System.Data.SqlClient are no longer part of Microsoft.AspNetCore.App.
+
+  **Preview 1 Known Issue** - the list of assemblies shipping in `Microsoft.AspNetCore.App` has not been finalized yet and will change before 3.0 RTM.
```
Comment (guardrex) [inline aspnetcore/migration/22-to-30.md]: I'd WARNING this one, too.

File: aspnetcore/migration/21-to-22.md [hunk 7]
```diff
@@ -142,6 +142,77 @@ Change the `FROM` lines in your *Dockerfile* to use the new image tags in the pr
 
 Visual Studio's **Auto build on browser request** experience doesn't function with the [IIS in-process hosting model](xref:fundamentals/servers/index#in-process-hosting-model). You must manually rebuild the project when using in-process hosting. Improvements to this experience are planned for a future release of Visual Studio.
 
+## Update logging code
+
+* Move logging provider initialization to `ConfigureServices`.
```
Comment (scottaddie) [inline aspnetcore/migration/21-to-22.md]:
```suggestion
* Move logging provider initialization to `Startup.ConfigureServices`.
```

File: aspnetcore/migration/21-to-22.md [hunk 8]
```diff
@@ -142,6 +142,77 @@ Change the `FROM` lines in your *Dockerfile* to use the new image tags in the pr
...
+  2.2 example:
+
+  ``` C#
+  public void ConfigureServices(IServiceCollection services)
+  {
+      services.AddLogging(builder => builder.AddConsole());
+  }
+  ```
+
+* Change filtering to use `AddLogging` builder.
```
Comment (scottaddie) [inline aspnetcore/migration/21-to-22.md]:
```suggestion
* Change filtering to use the `AddLogging` builder.
```

File: aspnetcore/migration/Logging-nonaspnetcore.md [hunk 9]
```diff
@@ -0,0 +1,69 @@
+---
+title: Migrate from Microsoft.Extensions.Logging 2.1 to 2.2 or 3.0
+author: pakrym
+description: This article outlines the common steps for migrating a non-ASP.NET Core application that uses Microsoft.Extensions.Logging 2.1 to Microsoft.Extensions.Logging 2.2 or 3.0.
```
Comment (guardrex) [inline aspnetcore/migration/Logging-nonaspnetcore.md]: Again ... "Learn ..." "Discover ..." ... this came down from some APEX guidance ages ago. I think while you were away we went thru and made a huge sweep to update all of the topics to that format.

File: aspnetcore/migration/21-to-22.md [hunk 10]
```diff
@@ -142,6 +142,77 @@ Change the `FROM` lines in your *Dockerfile* to use the new image tags in the pr
...
+      services.AddLogging(builder => builder
+          .AddConsole()
+          .AddFilter<ConsoleLoggerProvider>(category: null, level: LogLevel.Information)
+          // or
+          .AddFilter<ConsoleLoggerProvider>((category, level) => category == "A" || level == LogLevel.Critical)
+          );
+  }
+  ```
+
+* Change configuration loading to use `AddLogging` builder.
```
Comment (scottaddie) [inline aspnetcore/migration/21-to-22.md]:
```suggestion
* Change configuration loading to use the `AddLogging` builder.
```

File: aspnetcore/migration/Logging-nonaspnetcore.md [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+ms.author: pakrym
+ms.custom: mvc
+ms.date: 01/04/2019
+uid: migration/logging-nonaspnetcore
+---
+
```
Comment (guardrex) [inline aspnetcore/migration/Logging-nonaspnetcore.md]: We haven't been putting a blank line here.

File: aspnetcore/migration/22-to-30.md [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+ms.author: tdykstra
+ms.custom: mvc
+ms.date: 01/04/2019
+uid: migration/22-to-30
+---
+# Migrate from ASP.NET Core 2.2 to 3.0
+
+By [Scott Addie](https://github.com/scottaddie)
+
+This article explains how to update an existing ASP.NET Core 2.2 project to ASP.NET Core 3.0 Preview.
```
Comment (scottaddie) [inline aspnetcore/migration/22-to-30.md]:
```suggestion
This article explains how to update an existing ASP.NET Core 2.2 project to ASP.NET Core 3.0.
```

File: aspnetcore/migration/22-to-30.md [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+## Update the project file
+
+* Set the [Target Framework Moniker (TFM)](/dotnet/standard/frameworks#referring-to-frameworks) to `netcoreapp3.0`:
+
+  ```xml
+  <TargetFramework>netcoreapp3.0</TargetFramework>
+  ```
+
+* Remove the `<PackageReference>` to `Microsoft.AspNetCore.All` and `Microsoft.AspNetCore.App`.
```
Comment (scottaddie) [inline aspnetcore/migration/22-to-30.md]:
```suggestion
* Remove any `<PackageReference>` to `Microsoft.AspNetCore.All` or `Microsoft.AspNetCore.App`.
```

File: aspnetcore/migration/22-to-30.md [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+
+  **Preview 1 Known Issue** - workaround for [NuGet/Home #7342](https://github.com/NuGet/Home/issues/7342) - projects that do not start with `<Project Sdk="Microsoft.NET.Sdk.Web">` will get compiler or runtime errors due to missing Microsoft.AspNetCore.* assemblies. This is most often the case for test projects and class libraries. The workaround is to add the following to the *.csproj* file.
```
Comment (scottaddie) [inline aspnetcore/migration/22-to-30.md]:
```suggestion
  **Preview 1 Known Issue** - workaround for [NuGet/Home #7342](https://github.com/NuGet/Home/issues/7342) - projects that don't start with `<Project Sdk="Microsoft.NET.Sdk.Web">` will get compiler or runtime errors due to missing `Microsoft.AspNetCore.*` assemblies. This is most often the case for test projects and class libraries. The workaround is to add the following to the *.csproj* file.
```

File: aspnetcore/migration/Logging-nonaspnetcore.md [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+# Migrate from Microsoft.Extensions.Logging 2.1 to 2.2 or 3.0
+
+This article outlines the common steps for migrating a non-ASP.NET Core application that uses `Microsoft.Extensions.Logging` 2.1 to `Microsoft.Extensions.Logging` 2.2 or 3.0.
```
Comment (guardrex) [inline aspnetcore/migration/Logging-nonaspnetcore.md]: "app" favored over "application" ... I don't want to beat a :skull:🐴 tho. The last discussion was "where it makes sense;" but as I said during that discussion, that's pretty much 100% of the cases when "application" isn't part of a proper noun.

File: aspnetcore/migration/22-to-30.md [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+
+  ```xml
+  <ItemGroup>
+     <FrameworkReference Include="Microsoft.AspNetCore.App" />
+  </ItemGroup>
+  ```
+
+* Update the `Version` on remaining `<PackageReference>` elements for `Microsoft.AspNetCore.*` packages to the current preview (for example, 3.0.0-preview-18579-0053).
+
+  If there is no 3.0 version of a package, the package might have been deprecated in 3.0. Many of these are part of `Microsoft.AspNetCore.App` and should not be referenced individually anymore. For a preliminary list of packages no longer produced in 3.0, see [aspnet/AspNetCore #3756](https://github.com/aspnet/AspNetCore/issues/3756).
```
Comment (scottaddie) [inline aspnetcore/migration/22-to-30.md]:
```suggestion
  If there is no 3.0 version of a package, the package might have been deprecated in 3.0. Many of these are part of `Microsoft.AspNetCore.App` and shouldn't be referenced individually anymore. For a preliminary list of packages no longer produced in 3.0, see [aspnet/AspNetCore #3756](https://github.com/aspnet/AspNetCore/issues/3756).
```

File: aspnetcore/migration/Logging-nonaspnetcore.md [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+
+## 2.1 to 2.2
+
+Manually create `ServiceCollection` and call `AddLogging`.
```
Comment (guardrex) [inline aspnetcore/migration/Logging-nonaspnetcore.md]: He'll correct me if I'm wrong, but @scottaddie and I have been linking all of the first mentions of API to the API Browser. Still not possible yet for 2.2 AFAIK ... I think the API Browser is still broken.

File: aspnetcore/migration/22-to-30.md [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+
+* Some assemblies were removed from Microsoft.AspNetCore.App between 2.x and 3.0. You may need to add `<PackageReference>` items if you're using APIs from packages listed in [aspnet/AspNetCore #3755](https://github.com/aspnet/AspNetCore/issues/3755)
```
Comment (scottaddie) [inline aspnetcore/migration/22-to-30.md]:
```suggestion
* Some assemblies were removed from `Microsoft.AspNetCore.App` between 2.x and 3.0. You may need to add `<PackageReference>` items if you're using APIs from packages listed in [aspnet/AspNetCore #3755](https://github.com/aspnet/AspNetCore/issues/3755)
```

File: aspnetcore/migration/22-to-30.md [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+
+  For example, `Microsoft.EntityFrameworkCore` and `System.Data.SqlClient` are no longer part of Microsoft.AspNetCore.App.
```
Comment (scottaddie) [inline aspnetcore/migration/22-to-30.md]:
```suggestion
  For example, `Microsoft.EntityFrameworkCore` and `System.Data.SqlClient` are no longer part of `Microsoft.AspNetCore.App`.
```

File: aspnetcore/migration/22-to-30.md [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+
+  **Preview 1 Known Issue** - the list of assemblies shipping in `Microsoft.AspNetCore.App` has not been finalized yet and will change before 3.0 RTM.
```
Comment (scottaddie) [inline aspnetcore/migration/22-to-30.md]:
```suggestion
  **Preview 1 Known Issue** - the list of assemblies shipping in `Microsoft.AspNetCore.App` hasn't been finalized yet and will change before 3.0 RTM.
```

File: aspnetcore/migration/Logging-nonaspnetcore.md [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+
+2.1 example:
+
+``` C#
```
Comment (guardrex) [inline aspnetcore/migration/Logging-nonaspnetcore.md]:
```suggestion
```csharp
```

File: aspnetcore/migration/Logging-nonaspnetcore.md [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+using (var serviceProvider = serviceCollection.BuildServiceProvider())
+using (var loggerFactory = serviceProvider.GetService<ILoggerFactory>())
+{
+    // use loggerFactory
+}
+```
+
+## 2.1 to 3.0
+
+In 3.0 use `LoggingFactory.Create`.
```
Comment (scottaddie) [inline aspnetcore/migration/Logging-nonaspnetcore.md]:
```suggestion
In 3.0, use `LoggingFactory.Create`.
```

File: aspnetcore/migration/Logging-nonaspnetcore.md [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+{
+    loggerFactory.AddConsole();
+
+    // use loggerFactory
+}
+```
+
+2.2 example:
+
+``` C#
```
Comment (guardrex) [inline aspnetcore/migration/Logging-nonaspnetcore.md]:
```suggestion
```csharp
```

File: aspnetcore/toc.md [hunk 24]
```diff
@@ -392,6 +392,7 @@
 ## [Use a file watcher](xref:tutorials/dotnet-watch)
 
 # Migration
+## [2.2 to 3.0 ](xref:migration/22-to-30)
```
Comment (scottaddie) [inline aspnetcore/toc.md]:
```suggestion
## [2.2 to 3.0](xref:migration/22-to-30)
```

File: aspnetcore/migration/Logging-nonaspnetcore.md [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+}
+```
+
+## 2.1 to 3.0
+
+In 3.0 use `LoggingFactory.Create`.
+
+2.1 example:
+
+``` C#
```
Comment (guardrex) [inline aspnetcore/migration/Logging-nonaspnetcore.md]:
```suggestion
```csharp
```

File: aspnetcore/migration/Logging-nonaspnetcore.md [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+{
+    loggerFactory.AddConsole();
+
+    // use loggerFactory
+}
+```
+
+3.0 example:
+
+``` C#
```
Comment (guardrex) [inline aspnetcore/migration/Logging-nonaspnetcore.md]:
```suggestion
```csharp
```

File: aspnetcore/toc.md [hunk 27]
```diff
@@ -406,6 +407,7 @@
 ### [ClaimsPrincipal.Current](xref:migration/claimsprincipal-current)
 ### [Membership to Identity](xref:migration/proper-to-2x/membership-to-core-identity)
 ### [HTTP modules to middleware](xref:migration/http-modules)
+## [Logging (not ASP.NET Core)](xref:migration/logging-nonaspnetcore)
```
Comment (guardrex) [inline aspnetcore/toc.md]: "not ASP.NET Core" seems strange

Comment (guardrex) [reply aspnetcore/toc.md]: Nevermind ... I guess there's no better way.

File: aspnetcore/includes/net-core-prereqs-all-3.0.md [hunk 28]
```diff
@@ -0,0 +1,15 @@
...
+
+# [Visual Studio Code](#tab/visual-studio-code)
+
+[!INCLUDE[](~/includes/net-core-prereqs-vsc-3.0.md)]
+
+# [Visual Studio for Mac](#tab/visual-studio-mac)
+
+[!INCLUDE[](~/includes/net-core-prereqs-mac-3.0.md)]
+
+---
```
[the remark this answers is not in the document]
Comment (tdykstra) [reply aspnetcore/includes/net-core-prereqs-all-3.0.md]: I'm following what we have for 2.2.  Tabs are VS, VSC, VSMac.

[the remark this answers is not in the document]
Comment (guardrex) [reply aspnetcore/includes/net-core-prereqs-all-3.0.md]:
```
### [.NET Core CLI](#tab/netcore-cli)
```

File: aspnetcore/includes/net-core-prereqs-vs-3.0.md (hunk 3 above)
Comment (tdykstra) [reply aspnetcore/includes/net-core-prereqs-vs-3.0.md]: >Also, I suspect VS 2019 will be required when 3.0 releases. Do we have an issue to update prereqs accordingly?

I don't think we have been creating a dedicated issue for updating prereq includes each release. I would wait until VS 2019 becomes required; that'll be the sort of change that we won't need a reminder to change prereqs for.

**Comment by guardrex** [comment]
@tdykstra Another commit coming ... there are still "C#" code language markers and here's the tab for the CLI if you want to do (or use an existing) INCLUDE ...

```
# [.NET Core CLI](#tab/netcore-cli)
```

File: aspnetcore/migration/22-to-30.md [hunk 29]
```diff
@@ -0,0 +1,45 @@
...
+
+## Update the project file
+
+* Set the [Target Framework Moniker (TFM)](/dotnet/standard/frameworks#referring-to-frameworks) to `netcoreapp3.0`:
+
+  ```xml
+  <TargetFramework>netcoreapp3.0</TargetFramework>
+  ```
+
+* Remove any `<PackageReference>` to [Microsoft.AspNetCore.All](xref:fundamentals/metapackage)` or [Microsoft.AspNetCore.App](xref:fundamentals/metapackage-app).
```
Comment (guardrex) [inline aspnetcore/migration/22-to-30.md]: AFAIK (without much discussion on it), Rick, Scott, and I adopted a pattern while you were away. The links to the metapackage topics will say "metapackage" in the link, so I suggest (and there's a stray backtick there, too) ...

```suggestion
* Remove any `<PackageReference>` to [Microsoft.AspNetCore.All metapackage](xref:fundamentals/metapackage) or [Microsoft.AspNetCore.App metapackage](xref:fundamentals/metapackage-app).
```

File: aspnetcore/migration/22-to-30.md [hunk 30]
```diff
@@ -0,0 +1,45 @@
...
+
+  **Preview 1 Known Issue** - workaround for [NuGet/Home #7342](https://github.com/NuGet/Home/issues/7342) - projects that don't start with `<Project Sdk="Microsoft.NET.Sdk.Web">` will get compiler or runtime errors due to missing `Microsoft.AspNetCore.*` assemblies. This is most often the case for test projects and class libraries. The workaround is to add the following to the *.csproj* file.
```
Comment (guardrex) [inline aspnetcore/migration/22-to-30.md]: *ugh* ... imo, this doesn't look good and the future tense isn't necessary ... but ok.

File: aspnetcore/migration/22-to-30.md [hunk 31]
```diff
@@ -0,0 +1,45 @@
...
+
+* Update the `Version` on remaining `<PackageReference>` elements for `Microsoft.AspNetCore.*` packages to the current preview (for example, 3.0.0-preview-18579-0053).
+
+  If there is no 3.0 version of a package, the package might have been deprecated in 3.0. Many of these are part of `Microsoft.AspNetCore.App` and should not be referenced individually anymore. For a preliminary list of packages no longer produced in 3.0, see [aspnet/AspNetCore #3756](https://github.com/aspnet/AspNetCore/issues/3756).
+
+* Some assemblies were removed from `Microsoft.AspNetCore.App` between 2.x and 3.0. You may need to add `<PackageReference>` items if you're using APIs from packages listed in [aspnet/AspNetCore #3755](https://github.com/aspnet/AspNetCore/issues/3755)
+
+  For example, `Microsoft.EntityFrameworkCore` and `System.Data.SqlClient` are no longer part of `Microsoft.AspNetCore.App`.
+
+  **Preview 1 Known Issue** - the list of assemblies shipping in `Microsoft.AspNetCore.App` hasn't been finalized yet and will change before 3.0 RTM.
```
Comment (guardrex) [inline aspnetcore/migration/22-to-30.md]: Same here.

File: aspnetcore/migration/Logging-nonaspnetcore.md (hunk 11 above)
Comment (guardrex) [reply aspnetcore/migration/Logging-nonaspnetcore.md]: Blank line ... I actually went thru ages ago and made this consistent. Recommend that we don't put a blank line under the metadata or that we set up a PR to do it for every topic. *My obsessive-compulsive disorder is showing!* :smile: lol

**Comment by guardrex** [comment]
>  VSC and Mac use the CLI

I hope you'll confer tho on that point. IMO, it would be better to always have all of the tabs. There are devs who like CLI things. It's not obvious to those devs what these IDEs are using. Anyway ... it's just a nit.

**Comment by tdykstra** [comment]
@guardrex I addressed some of the nits, though not quite exactly the way you had in mind. 😄 

> I hope you'll confer tho on that point.

Will do

**Pull request merged**